### DISCUSSION TITLE
Plugin support?

### DISCUSSION THREAD

**Discussion by ukstv** [pnpm/pnpm#3350] [Ideas]
First, thank you for all the hard work on pnpm. IMO it is _the best_ node package manager out there.

I am working almost exclusively on monorepos these days. I found it useful to perform commands on (this package and all the dependencies in a workspace). Like, build (this and the dependencies) or list (this and the dependencies). Even written some clunky JS helper for that. Would be cool to declare that in pnpmfile or something else, and use it as vanilla pnpm commands. I wonder, if there is a plan for pnpm to support some sort of plugin system to add commands directly to PNPM.

**Comment by zkochan** [comment]
Yes, we would love that. And there is one candidate already: https://github.com/pnpm/pnpm/pull/3302

**Reply by ukstv** [reply]
Based on discussion there, my understanding was it will be a separate command, like `pnpm-graph` or something, not `pnpm graph`. I was asking more about how to add this as `pnpm graph` callable even if it is provided in a separate package.

**Reply by ukstv** [reply]
Or maybe general preferred idea is "get your CLI separate from pnpm"?

**Reply by zkochan** [reply]
Yes, I understand what you mean.

**Reply by zkochan** [reply]
Well, I am not sure. There are pros and cons. pnpm is already built on internal plugins, so we have an API we could share. Whether it is a good idea, I don't know. Pluggable systems are powerful but they also add complexity.

**Reply by ukstv** [reply]
Understood.

**Reply by ScottAwesome** [reply]
If there is a style of good conventions and the plugin API isn't changing too often, I don't think exposing it would be a terrible idea. I think having a pluggable package manager would simplify a lot of complexity around how tools develop their CLIs in the first place (if you can just hook into the package manager its a lot less complex) and good ideas for command line interfaces could be surfaced in a central location

I understand the complexity is also on the maintainer side though, as I can imagine this would open you up to having to deal with a lot more feature requests

**Comment by Kadeluxe** [comment]
I wonder, is it possible to hook some pnpm «processes»? 
Like hook into the resolution step to modify how pnpm resolves packages, etc.
I guess currently there is no such option, nor it's planned for the future?
